### PR TITLE
Do not automatically set warning flags if buildtype is 'plain'

### PR DESCRIPTION
**Pull request by ximion**
Like the title says.

The `no_warn_args` flag seems to be primarily relevant for Vala, so I didn't use ot touch that one even if we are in 'plain' mode.

This PR also includes a minimal change to make the warning handling when LDC a bit better (warning level 3 would have had *less* warnings than warning level 2 before).

See https://github.com/mesonbuild/meson/pull/1214 for reference.

### CODE REVIEW

File: mesonbuild/compilers.py [hunk 1]
```diff
@@ -1758,8 +1758,8 @@ def get_include_args(self, path, is_system):
         return ['-I' + path]
 
     def get_warn_args(self, level):
-        if level == '2':
-            return ['-wi']
+        if level == '2' or level == '3':
```
Comment (nirbheek) [inline mesonbuild/compilers.py]: Do you think you can find out if the same arguments also work with other D compilers and apply this change there too?

Comment (ximion) [inline mesonbuild/compilers.py]: GDC is covered, DMD's help says:
```
-w             warnings as errors (compilation will halt)
-wi            warnings as messages (compilation will continue)
```
while LDC says:
```
-w                                      - Enable warnings
-wi                                     - Enable informational warnings
```

I think there is a bug in LDC if -w is really -Werror, either in the docs or in its implementation. In any case, it looks like I need to clarify this with upstream and then adjust Meson again. :-/

**Comment by ximion** [comment]
The D stuff is updated now too, with information from upstream. So, this change should be mergeable (unless you have a better idea on how to handle warning flags when buildtype is 'plain').

**Comment by ximion** [comment]
I added another fix when compiling D stuff with LDC... Unfortunately Github doesn't easily let me create multiple pull requests, so this all ended up in this one... Meh.

File: mesonbuild/compilers.py [hunk 2]
```diff
@@ -1639,7 +1639,8 @@ def get_std_shared_lib_link_args(self):
         return ['-shared']
 
     def get_soname_args(self, prefix, shlib_name, suffix, path, soversion, is_shared_module):
-        return []
+        # FIXME: Make this work for Windows, MacOS and cross-compiling
+        return get_gcc_soname_args(GCC_STANDARD, prefix, shlib_name, suffix, path, soversion, is_shared_module)
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E501] line too long (111 > 79 characters)

File: mesonbuild/backend/backends.py [hunk 3]
```diff
@@ -329,7 +329,7 @@ def generate_basic_compiler_args(self, target, compiler, no_warn_args=False):
         commands += compiler.get_always_args()
         if no_warn_args:
             commands += compiler.get_no_warn_args()
-        else:
+        elif self.environment.coredata.get_builtin_option('buildtype') != 'plain':
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[flake8]__
* [E501] line too long (82 > 79 characters)

File: mesonbuild/compilers.py (hunk 2 above)
Comment (jpakkane) [inline mesonbuild/compilers.py]: We do not have, nor will there ever be, an 80 character line limit. EVER!

Comment (ximion) [inline mesonbuild/compilers.py]: This line is copied almost verbatim, so we would need to adjust a lot of other places to fit that requirement as well ;-)

**Comment by ximion** [comment]
Would be nice to get these changes merged, since compiling any more complex D code (which includes libraries or is Vibe.d) is impossible with the current Meson.

File: mesonbuild/compilers.py (hunk 2 above)
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: yep, just rebase on top of master

File: mesonbuild/backend/backends.py [hunk 4]
```diff
@@ -364,7 +364,7 @@ def build_target_link_arguments(self, compiler, deps):
             if not isinstance(d, (build.StaticLibrary, build.SharedLibrary)):
                 raise RuntimeError('Tried to link with a non-library target "%s".' % d.get_basename())
             if isinstance(compiler, compilers.LLVMDCompiler):
-                args.extend(['-L', self.get_target_filename_for_linking(d)])
+                args += ['-L' + self.get_target_filename_for_linking(d)]
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: nice catch

File: mesonbuild/backend/backends.py (hunk 3 above)
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: is it really necessary for this PR? I think it needs more discussion.

Comment (ximion) [inline mesonbuild/backend/backends.py]: This is resolving the problem that 'plain' isn't actually 'plain' in Meson but adds some warning flags unconditionally.
On distributions, if we e.g. want to enable only specific warnings (e.g. implicit-function-declaration) and make them fatal, this behavior would be an issue.

It's a separate bug though and would be in a separate PR if Github would let me do that.

File: mesonbuild/compilers.py [hunk 5]
```diff
@@ -1635,7 +1635,8 @@ def get_std_shared_lib_link_args(self):
...
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E501] line too long (111 > 100 characters)

Comment (QuLogic) [inline mesonbuild/compilers.py]: I guess you need to set a longer limit...
